File: typescript/src/tools/MCPTool/MCPTool.ts
import { Tool, ToolUseContext } from '../../infrastructure/tools/Tool';
import { McpClientInstance } from '../../services/mcp/client';
import { buildMcpToolName } from '../../services/mcp/mcpStringUtils';

/**
 * MCP 代理工具创建工厂 (透明代理模式)
 * 
 * 核心原理：
 * 这个函数负责把远端 MCP Server 返回的工具定义（Schema），
 * 包装成 Claude Code 引擎能够直接理解的本地 `Tool<T, string>` 接口。
 * 
 * 对大模型（QueryEngine）而言，它根本不知道它在调用一个跑在其他进程（甚至可能是远程服务器上）的工具。
 * 它只管传入参数，然后获取返回的字符串。这实现了极大的解耦！
 *
 * @param serverName MCP 服务器的名称
 * @param mcpClient 已建立连接的 MCP 客户端实例
 * @param toolDef 远端返回的工具定义（包含工具名、描述、输入参数的 Schema）
 * @returns 一个符合 Claude Code 引擎标准的 Tool 接口实例
 */
export function createMcpTool(
  serverName: string,
  mcpClient: McpClientInstance,
  toolDef: any
): Tool<any, string> {
  // 使用 mcpStringUtils.ts 中的方法为工具名称打上包装前缀
  // 防止不同插件中出现重名的工具
  const proxyName = buildMcpToolName(serverName, toolDef.name);

  return {
    name: proxyName,
    // 在描述前面加上插件来源，有助于大模型在思考时明白这个工具的背景
    description: `(MCP 插件来自 ${serverName}) ${toolDef.description || ''}`,
    // 直接透传远端的参数定义 Schema 给大模型
    inputSchema: toolDef.inputSchema,
    
    // 大模型决定调用该工具时，引擎会执行这个方法
    execute: async (args: any, context: ToolUseContext): Promise<string> => {
      console.log(`[MCPTool] 转发请求到 ${serverName} 插件的工具 ${toolDef.name}...`);
      
      try {
        // 通过基于 stdio 的 JSON-RPC 连接，将调用请求序列化后发送给远端子进程
        const response = await mcpClient.client.callTool({
          name: toolDef.name,
          arguments: args,
        });

        // 解析远端返回的数据结构
        // 根据 MCP 协议规范，返回值中的 content 是一个包含了不同媒体类型的数组
        if (response && response.content && Array.isArray(response.content)) {
          // 我们这里提取出所有的文本类型内容，并将其拼接成字符串返回给大模型
          return response.content
            .filter((c: any) => c.type === 'text')
            .map((c: any) => c.text)
            .join('\n');
        }
        
        // 如果返回的格式不标准，直接将其序列化为字符串兜底
        return JSON.stringify(response);
      } catch (error: any) {
        console.error(`[MCPTool] 调用插件 ${serverName} 失败:`, error);
        // 插件挂了或者报错，优雅地将错误信息返回给大模型，大模型看到错误后可能会尝试自我修正
        return `插件执行失败: ${error.message}`;
      }
    },
  };
}

File: python/src/mini_cc/tools/mcp_tool.py
import json
import asyncio
from pydantic import BaseModel, Field
from typing import Dict, Any, Optional

from .base import BaseTool

class MCPToolArgs(BaseModel):
    # MCP 工具的参数是动态的，所以我们用 dict 来接收
    # 实际上，注册到大模型时，我们会直接把 MCP Server 返回的 JSON Schema 喂给大模型
    args: Dict[str, Any] = Field(default_factory=dict, description="传递给 MCP 插件的参数字典")

class MCPTool(BaseTool):
    """
    透明代理工具 (Model Context Protocol)
    这是一个特殊的 Tool。它不是在执行本地 Python 代码，
    而是将大模型的请求通过 stdio 或 HTTP 转发给另外一个独立的 MCP Server 进程。
    """
    def __init__(self, name: str, description: str, server_command: str, args_schema: dict):
        self.name = name
        self.description = description
        self.server_command = server_command
        self._custom_schema = args_schema
        # 默认的参数校验模型，虽然实际上我们会重写 to_openai_schema
        self.args_schema = MCPToolArgs
        
    def to_openai_schema(self) -> dict:
        """
        重写 schema 生成逻辑，直接使用从 MCP Server 读来的真实 Schema
        """
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": self._custom_schema
            }
        }
        
    async def execute(self, **kwargs) -> str:
        """
        核心逻辑：当大模型想调用这个 MCP 工具时，
        我们启动 MCP Server 进程，把 JSON 发给它，并读取它的 stdout 响应。
        """
        try:
            # 构造符合 MCP 协议规范的 JSON-RPC 请求
            request = {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "tools/call",
                "params": {
                    "name": self.name,
                    "arguments": kwargs
                }
            }
            
            req_str = json.dumps(request) + "\n"
            
            # 通过 asyncio.create_subprocess_shell 启动基于 stdio 的 MCP Server
            process = await asyncio.create_subprocess_shell(
                self.server_command,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE
            )
            
            # 向子进程发送请求并等待响应
            stdout, stderr = await process.communicate(input=req_str.encode('utf-8'))
            
            if process.returncode != 0:
                err_msg = stderr.decode('utf-8').strip()
                return f"MCP 插件执行失败 (退出码 {process.returncode}): {err_msg}"
                
            # 解析 MCP Server 返回的 JSON-RPC 响应
            resp_str = stdout.decode('utf-8').strip()
            
            try:
                # 有些 MCP Server 可能会混杂一些普通的 log 打印，我们需要找到最后那行合法的 JSON
                lines = [line for line in resp_str.split('\n') if line.strip().startswith('{')]
                if not lines:
                    return f"MCP 插件返回格式错误: {resp_str}"
                    
                resp_data = json.loads(lines[-1])
                
                if "error" in resp_data:
                    return f"MCP 插件业务错误: {resp_data['error'].get('message', '未知错误')}"
                    
                # 提取返回内容
                result_content = resp_data.get("result", {}).get("content", [])
                if not result_content:
                    return "MCP 插件执行成功，但没有返回内容。"
                    
                # 将 MCP 的 content 数组（可能包含 text/image）拼接成字符串发给大模型
                texts = [item.get("text", "") for item in result_content if item.get("type") == "text"]
                return "\n".join(texts)
                
            except json.JSONDecodeError:
                return f"无法解析 MCP 插件响应: {resp_str}"
                
        except Exception as e:
            return f"调用 MCP 插件时发生系统错误: {str(e)}"
